### ISSUE TITLE
Apply global config and override for certain files

### ISSUE THREAD

**Issue by OliverJAsh** [editorconfig/editorconfig#220]
Given:

```
; .editorconfig
[**.foo]
indent_style = tab
```

```
; project/.editorconfig
[*]
indent_style = space
```

When editing `project/bar.foo`, EditorConfig will apply `indent_style = space`. I want `indent_style = tab`.

I want to set a global configuration for certain file types, e.g. `.foo`, which should override any "closer" configurations.

Could EditorConfig order styles by specificity? Or is there another way to do this?

**Comment by xuhdev** [comment]
I don't think there is a way to do this right now but we can discuss the possibility. One way to do it may be adding an "include" directive, then you can always include the first piece of code in your second project .editorconfig file.

**Comment by treyhunner** [comment]
There is no way to do this right now.  You would typically want to include `[**.foo]` in your local project's `.editorconfig` file if needed or *not* use `*` if the configuration should not be universal.

@OliverJAsh what is your use case for this?

**Comment by xuhdev** [comment]
I revisit your problem again and I think your case may be a bit anti-pattern. `.editorconfig` is supposed to define consistent coding style for **each project**. If you use your personal preference for some files in the project, how do you make sure other people use the same coding style? They don't have access to your `~/.editorconfig` file any way.

**Comment by OliverJAsh** [comment]
My use case: I have a global file, `project-name.sublime-project`, which lives in each of my Git directories but I do not commit it to the repo. It is in that folder for my convenience.

I want control of the config for this file.

This might be an edge case.

**Comment by treyhunner** [comment]
@OliverJAsh there is no override built-in to the EditorConfig core libraries.

As a workaround, you could customize the Sublime Text plugin code to specify a hard-coded set of properties (probably just an empty list to use Sublime defaults) if the file name ends in `.sublime-project` ([see code here][code]).  I'm not very familiar with the way Sublime Text works, but if that might be a useful feature for others maybe a @sindresorhus would consider a pull request.

[code]: https://github.com/sindresorhus/editorconfig-sublime/blob/master/EditorConfig.py#L51

**Comment by sindresorhus** [comment]
:-1:

**Comment by selfthinker** [comment]
@OliverJAsh, `project-name.sublime-project` doesn't need to live in each project's folder, in can go anywhere in the file system (e.g. parent folder or "projects" folder, etc).
You might know this already and have chosen deliberately to put it into your project folders. But in case you didn't know it, it's not necessary.

**Comment by OliverJAsh** [comment]
I just ran into this again. Here is what I think is a good real world use case.

I have set my global EditorConfig (`~/.editorconfig`) to trim trailing whitespace from all files except `*.diff`, because trailing whitespace matters in those files ([if it's trimmed it will corrupt the patch](https://stackoverflow.com/questions/50258565/git-editing-hunks-fails-when-file-has-other-hunks/50275053#50275053)):

``` ini
root = true

[*]
trim_trailing_whitespace = true

[*.diff]
trim_trailing_whitespace = false
```

Then, a project I work on defines its own settings:

``` ini
[*]
trim_trailing_whitespace = true
```

This project setting will override and therefore break my global preference for `*.diff` files.

This means that when editing Git hunks (`git add -p` and then `e` for edit) for this project, which opens the file `$PROJECT/.git/addp-hunk-edit.diff`, the trailing whitespace will be trimmed, therefore corrupting my patch.

Of course I can redeclare the `*.diff` setting in the project specific settings, but that would be redundant for most users, who do not edit Git hunks and therefore don't have this problem.

I believe this needs to be a global setting because it's not something that is common to the project and the team I work on, but the (temporary) file lives happens to live inside the project.

**Comment by slifty** [comment]
Another very important use case here which is not specific to file types: accessibility

[This thread on tabs / spaces](https://www.reddit.com/r/javascript/comments/c8drjo/nobody_talks_about_the_real_reason_to_use_tabs/) explains how tabs are critical for certain users who need to adjust the width of indentation to accommodate their visual accessibility needs.

Editorconfig was cited as a way to set a default width on general services like github, but users need to be able to globally override on their local devices based on their accessibility needs regardless of the specified indent_size.

It seems that the global settings get overridden by project specific settings.  It would be great to allow users with accessibility concerns to override as part of the editorconfig spec.

**Comment by Zsar** [comment]
Same use case as [OliverJAshes](https://github.com/editorconfig/editorconfig/issues/220#issuecomment-388073847) with different file type:

   	root = true

	[*]
	trim_trailing_whitespace = true

	[**.md]
	trim_trailing_whitespace = false # forced linebreak: two trailing whitespaces

In .md files trailing whitespaces are a formatting hint and must be retained.

It is currently not possible to express a file pattern "all but .md or .diff files" (nor have I found a feature request for one; #228 and #261 look similar but seem to aim at excluding files from _any and all_ .editorconfig treatment).

tl;dr: This feature is interesting wherever the list of "special" file types is far shorter than the list of uniformly handled file types in a project. Aka "the general case", I dare say.

**Comment by kitsunekyo** [comment]
necroing this issue as I have a similar issue where the repo i am working in has set a `indent_size = 8` in a `.editorconfig` that is committed to the repo.

for accessibility i need a much smaller `indent_size`. this only affects rendering of tab width so it is entirely personal and highly subjective. at the moment, the only way around this is to disable editorconfig handling in my IDE completely.

i was initially searching for something similar to how `.env` files are inherited: `.editorconfig.local` could be gitignored and amended for personal preference.

**Comment by cxw42** [comment]
> rule priority

https://spec.editorconfig.org/#file-processing

> pairs defined in the section that comes later in the .editorconfig file take precedence over pairs defined in the section that comes earlier in the same .editorconfig file. If multiple EditorConfig files have matching sections, the pairs from the closer EditorConfig file are read last, so pairs in closer files take precedence.

> the only way for a person with visual impairment to meet their personal needs would be to use an IDE that doesn't support editorconfig :(

For some editors (e.g., vim), you can override only specific editorconfig settings in code.  However, there is no universal mechanism I know of for such overrides.

**Comment by xuhdev** [comment]
> 
> > the only way for a person with visual impairment to meet their personal needs would be to use an IDE that doesn't support editorconfig :(
> 
> For some editors (e.g., vim), you can override only specific editorconfig settings in code. However, there is no universal mechanism I know of for such overrides.

I think it's better addressed by editors. EditorConfig is meant to set how the edited files will become (as to ensure consistent code styles across editors for a team). For tweaks about individual visual preferences (e.g., an editor should display one leading space as two), I believe this is out of scope of the project.

**Comment by cxw42** [comment]
>  I can redeclare the *.diff setting in the project specific settings, 

This is in fact the recommended way to go.  That way everyone who works on that project will have the same settings.  (If a project maintainer isn't willing to accept additions to `.editorconfig`, there's not much we can do about that, unfortunately :( .)

>  repo ... has set a indent_size = 8 ... i need a much smaller ... only affects rendering of tab width

Sorry you're running into this!  Am I correct in thinking that the repo has `indent_style = tab`?  If not, EditorConfig really can't do much.

All else being equal, changes that don't affect disk contents are likely to be out of scope for the EditorConfig core, as @xuhdev mentioned.  However, the plugin for your editor might include an extension mechanism.  I see you mentioned "IDE" @kitsunekyo but I don't see a mention of which IDE --- I can try to give you a specific example if you let me know!  

(non-IDE example: editorconfig-vim has an [`AddNewHook()`](https://github.com/editorconfig/editorconfig-vim/blob/6a58b7c11f79c0e1d0f20533b3f42f2a11490cf8/autoload/editorconfig.vim#L37) to register a function that will be called after all other editorconfig processing.  You could implement indent_width changes in a hook.)

> Could EditorConfig order styles by specificity? 

That is being discussed some in #236 as well.  (See also #415, where I'm hoping to get some input on where the project might go in the future.)
